agent/agent.go

```diff
@@ -0,0 +1,265 @@
+}
+
+// Returns full profile for agent.
+func (a *Agent) GetFullProfile() map[string]interface{} {
+	return map[string]interface{}{
+		"paw": a.paw,
+		"server": a.server,
+		"group": a.group,
+		"host": a.host,
+		"username": a.username,
+		"architecture": a.architecture,
+		"platform": a.platform,
+		"location": a.location,
+		"pid": a.pid,
+		"ppid": a.ppid,
+		"executors": a.executors,
+		"privilege": a.privilege,
+		"exe_name": a.exe_name,
+	}
+}
+
+// Return minimal subset of agent profile.
+func (a *Agent) GetTrimmedProfile() map[string]interface{} {
+	return map[string]interface{}{
+		"paw": a.paw,
+		"server": a.server,
+		"platform": a.platform,
+		"host": a.host,
+	}
+}
+
+// Pings C2 for instructions and returns them.
+func (a *Agent) Beacon() map[string]interface{} {
+	var beacon map[string]interface{}
+	profile := a.GetFullProfile()
+	response := a.beaconContact.GetBeaconBytes(profile)
+	if response != nil {
+		beacon = processBeacon(response)
+	} else {
+		output.VerbosePrint("[-] beacon: DEAD")
+	}
+	return beacon
+}
+
+// Converts the given data into a beacon with instructions.
+func processBeacon(data []byte) map[string]interface{} {
+	var beacon map[string]interface{}
+	if err := json.Unmarshal(data, &beacon); err != nil {
+		output.VerbosePrint(fmt.Sprintf("[-] Malformed beacon received: %s", err.Error()))
+	} else {
+		var commands interface{}
+		if err := json.Unmarshal([]byte(beacon["instructions"].(string)), &commands); err != nil {
+			output.VerbosePrint(fmt.Sprintf("[-] Malformed beacon instructions received: %s", err.Error()))
+		} else {
+			output.VerbosePrint("[+] beacon: ALIVE")
+			beacon["sleep"] = int(beacon["sleep"].(float64))
+			beacon["watchdog"] = int(beacon["watchdog"].(float64))
+			beacon["instructions"] = commands
+		}
+	}
+	return beacon
+}
+
+func (a *Agent) Heartbeat() {
+	// TODO add any heartbeat functionality here.
+}
+
+func (a *Agent) Terminate() {
+	// Add any cleanup/termination functionality here.
+	output.VerbosePrint("[*] Terminating Sandcat Agent... goodbye.")
+}
+
+// Runs a single instruction and send results.
+func (a *Agent) RunInstruction(command map[string]interface{}, payloads []string) {
+	timeout := int(command["timeout"].(float64))
+	result := make(map[string]interface{})
+	commandOutput, status, pid := execute.RunCommand(command["command"].(string), payloads, command["executor"].(string), timeout)
+	result["id"] = command["id"]
+	result["output"] = commandOutput
+	result["status"] = status
+	result["pid"] = pid
+ 	a.beaconContact.SendExecutionResults(a.GetTrimmedProfile(), result)
+}
+
+// Sets the C2 communication channels for the agent according to the specified C2 configuration map.
+// Will default to HTTP if requested C2 is not available or its requirements aren't met. If HTTP is
+// not available or if no communication channels are available, an error will be returned.
+func (a *Agent) SetCommunicationChannels(c2Config map[string]string) error {
+	var err error
```

Comment:
This function can be cleaned up quite a bit.  First do the check for the desired C2, then only set the comms to HTTP if that fails